### PR TITLE
Add hcfoldMap, hctraverse etc.

### PR DESCRIPTION
**Pull request by phadej**
This would complete Applicative-like story. 

Maybe it worth adding `HTraverse` (or method to `HSequence` ?)

--

The primary motivation is `hcfoldMap` (which you *cannot* implement generically in terms of `hcollapse`!)

---

@kosmikus are you ok if we add these functions?

### CODE REVIEW

**Comment by phadej** [comment]
Questions:
- do we need `htraverse` (non-constrained). Yes for completeness, probably no in practice.
- do we need `hcfoldMap` in a class? (same with `hsequence'`) Should they be derived?

File: src/Generics/SOP/Classes.hs [hunk 1]
```diff
@@ -374,6 +380,26 @@ class HCollapse (h :: (k -> *) -> (l -> *)) where
...
 
+-- | A generalization of 'Data.Foldable.foldMap'.
+class HFoldMap (h :: (k -> *) -> (l -> *)) where
+
+  -- | Corresponds to 'Data.Foldable.foldMap'.
+  --
+  -- /Instances:/
+  --
+  -- @
+  -- ...
```
Comment (kosmikus) [inline src/Generics/SOP/Classes.hs]: These should be added.

File: src/Generics/SOP/Classes.hs [hunk 2]
```diff
@@ -374,6 +380,26 @@ class HCollapse (h :: (k -> *) -> (l -> *)) where
...
+  -- @
+  hcfoldMap :: (AllN h c xs, Monoid m) => proxy c -> (forall a. c a => f a -> m) -> h f xs -> m
+  hcfoldMap p f = unK . hctraverse_ p (K . f)
+
+  -- | Corresponds to 'Data.Foldable.traverse_'.
+  hctraverse_ :: (AllN h c xs, Applicative g) => proxy c -> (forall a. c a => f a -> g ()) -> h f xs -> g ()
+
+-- | Flipped version of 'hcfor_'
```
Comment (kosmikus) [inline src/Generics/SOP/Classes.hs]: Should this read "of `hctraverse_`"?

File: src/Generics/SOP/Classes.hs [hunk 3]
```diff
@@ -374,6 +380,26 @@ class HCollapse (h :: (k -> *) -> (l -> *)) where
...
+  -- /Instances:/
+  --
+  -- @
+  -- ...
+  -- @
+  hcfoldMap :: (AllN h c xs, Monoid m) => proxy c -> (forall a. c a => f a -> m) -> h f xs -> m
+  hcfoldMap p f = unK . hctraverse_ p (K . f)
+
+  -- | Corresponds to 'Data.Foldable.traverse_'.
+  hctraverse_ :: (AllN h c xs, Applicative g) => proxy c -> (forall a. c a => f a -> g ()) -> h f xs -> g ()
```
Comment (kosmikus) [inline src/Generics/SOP/Classes.hs]: What about `hcfoldr`? Should a connection be established to the catamorphisms?

**Comment by kosmikus** [comment]
Regarding your other questions:

- Yes, I'd always add the unconstrained versions of the functions for completeness.
- I think the assumption of `generics-sop` so far is that the only types we're interested in are `NP`, `POP`, `NS` and `SOP`, and so all the `H*` classes have at most four instances. As long as we wouldn't want a deviating definition for one of these, it's completely ok to define functions outside of the classes.

**Comment by kosmikus** [comment]
Oh, and could you please tell me once again why `hcfoldMap` is not implementable in terms of `hcollapse`?

**Comment by phadej** [comment]
hcollapse :: ... -> k, [k] or [[k]] depending on the structure (NS, NP and SOP, POP respectively). hcfoldMap is always -> m. From my limited pov, hctraverse_ is more useful variant, maybe because monadic DSLs are more common. But hcfoldMap ja hctraverse_ can be defined in terms of each other.

**Comment by phadej** [comment]
@kosmikus I added `htraverse_` and `htraverse'` as derived functions (`hpure` is class member), so there is `AllN h Top cs` - trivial constraint. Or is it better to have these functions as members?

Otherwise I think I addressed all your comments & concerns.

**Comment by kosmikus** [comment]
@phadej Defining the non-constrained variants in terms of the constrained variants by instantiating the constraint to `Top` is currently not idiomatic, i.e., not consistent with the rest of the package. These should instead just have `SListIN` constraints.

I do want to utlimately reorganise things such that `SListI ~ All Top`, but last time I tried, this requires abandoning pre-ghc-8 compatibility for certain technical reasons (IIRC, something to do with undecidable superclasses).

**Comment by kosmikus** [comment]
@phadej I have now merged this, but also made the change to implement the unconstrained variants directly. I also cleaned up a bit (comments, exports). There's still the question whether more of the unconstrained variants or specialized versions should be defined, but it's of somewhat lesser importance.

**Comment by phadej** [comment]
@kosmikus great. Things are easier to add, than to remove. So let's see if we'll need them :)

**Pull request merged**